keys/signer_secp256r1.go

```diff
@@ -0,0 +1,126 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+package keys
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+
+	"github.com/coinbase/rosetta-sdk-go/asserter"
+	"github.com/coinbase/rosetta-sdk-go/types"
+)
+
+// SignerSecp256r1 is initialized from a keypair
+type SignerSecp256r1 struct {
+	KeyPair *KeyPair
+}
+
+// The Ecdsa signature is the couple (r, s), both r and s are 32 bytes
+const (
+	EcdsaRLen = 32
+	EcdsaSLen = 32
+)
+
+var _ Signer = (*SignerSecp256r1)(nil)
+
+// PublicKey returns the PublicKey of the signer
+func (s *SignerSecp256r1) PublicKey() *types.PublicKey {
+	return s.KeyPair.PublicKey
+}
+
+// Sign arbitrary payloads using a KeyPair
+func (s *SignerSecp256r1) Sign(
+	payload *types.SigningPayload,
+	sigType types.SignatureType,
+) (*types.Signature, error) {
+	if err := s.KeyPair.IsValid(); err != nil {
+		return nil, err
+	}
+
+	if !(payload.SignatureType == sigType || payload.SignatureType == "") {
+		return nil, fmt.Errorf(
+			"%w: %v",
+			ErrSignUnsupportedPayloadSignatureType,
+			payload.SignatureType,
+		)
+	}
+
+	if sigType != types.Ecdsa {
+		return nil, fmt.Errorf(
+			"%w: expected %v but got %v",
+			ErrSignUnsupportedSignatureType,
+			types.Ecdsa,
+			sigType,
+		)
+	}
+
+	crv := elliptic.P256()
+	x, y := crv.ScalarBaseMult(s.KeyPair.PrivateKey)
+	pubKey := ecdsa.PublicKey{X: x, Y: y, Curve: crv}
+	privKey := ecdsa.PrivateKey{
+		PublicKey: pubKey,
+		D:         new(big.Int).SetBytes(s.KeyPair.PrivateKey),
+	}
+
+	sigR, sigS, err := ecdsa.Sign(rand.Reader, &privKey, payload.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrSignFailed, err.Error())
+	}
+	sig := sigR.Bytes()
+	sig = append(sig, sigS.Bytes()...)
+
+	return &types.Signature{
+		SigningPayload: payload,
+		PublicKey:      s.KeyPair.PublicKey,
+		SignatureType:  payload.SignatureType,
+		Bytes:          sig,
+	}, nil
+}
+
+// Verify verifies a Signature, by checking the validity of a Signature,
+// the SigningPayload, and the PublicKey of the Signature.
+func (s *SignerSecp256r1) Verify(signature *types.Signature) error {
+	if signature.SignatureType != types.Ecdsa {
+		return fmt.Errorf(
+			"%w: expected %v but got %v",
+			ErrVerifyUnsupportedSignatureType,
+			types.Ecdsa,
+			signature.SignatureType,
+		)
+	}
+
+	if err := asserter.Signatures([]*types.Signature{signature}); err != nil {
+		return err
+	}
+
+	message := signature.SigningPayload.Bytes
+	sig := signature.Bytes
+
+	x, y := elliptic.Unmarshal(elliptic.P256(), signature.PublicKey.Bytes)
+	publicKey := ecdsa.PublicKey{X: x, Y: y, Curve: elliptic.P256()}
```

Comment:
Check that the key is not the identity element.

Reply:
fixed

Reply:
This checks that the points are on the curve which is excellent but need to also check for the identity element which is when x and y are zero.

Reply:
Per above, IsOnCurve checks this